### ISSUE TITLE
Slaves crash on win64 with error "Not properly terminated"

### ISSUE THREAD

**Issue by rekcahpassyla** [pytest-dev/pytest-xdist#70]
Following on to #68 - after setting the `PYTEST_DEBUG` environment variable, I get the following output when I run using more than 1 process for example:

```
py.test -v  -n 3 -m regression --max-slave-restart=0 [redacted]
```

```

Slave restarting disabled
      pytest_testnodedown [hook]
          node: <SlaveController gw1>
          error: Not properly terminated
[gw1] node down: Not properly terminated
      finish pytest_testnodedown --> [] [hook]
      pytest_runtest_logreport [hook]
          report: <TestReport 'redacted.py::Redacted::test_eta2' when='???' outcome='failed'>
        pytest_report_teststatus [hook]
            report: <TestReport 'redacted.py::Redacted::test_eta2' when='???' outcome='failed'>
        finish pytest_report_teststatus --> ('failed', 'f', 'FAILED') [hook]
[gw1] FAILED redacted.py::Redacted::test_eta2       finish pytest_runtest_logreport --> [] [hook]

```

The set of tests that fail is different upon different runs, but all failures are with the above error (`Not properly terminated`). 

versions:

``` python
pytest                    2.9.1                    py27_0    
pytest-xdist              1.13.1                   py27_0    
```

**Comment by nicoddemus** [comment]
*Not properly terminated* means that the worker crashed, usually with a segfault. I assume the tests pass without `pytest-xdist`, right?

I suggest you try to run using [pytest-random](https://pypi.python.org/pypi/pytest-random) without xdist to see if you have the same behavior; if so, it means that the fact that xdist os running tests  in an unpredictable order is exposing some problem with your test suite, for example sharing some global state which only is a problem when tests run in a particular order (which is different from the normal order when executing pytest without xdist).

**Comment by rekcahpassyla** [comment]
Thanks- just tried that: ten times in a row, single-threaded, randomised: passed every time..

**Comment by nicoddemus** [comment]
Another idea: install [pytest-faulthandler](https://pypi.python.org/pypi/pytest-faulthandler) and see if that produces more useful output.

**Comment by rekcahpassyla** [comment]
Just tried with `pytest-faulthandler` - no more output than before, I'm afraid. 

```
================================== FAILURES ===================================
_ redacted.py __
[gw2] win32 -- Python 2.7.11 C:\dev\bin\Anaconda\python.exe
Slave 'gw2' crashed while running 'redacted::test_snt'
_ redacted __
[gw1] win32 -- Python 2.7.11 C:\dev\bin\Anaconda\python.exe
Slave 'gw1' crashed while running 'redacted::test_eta2'
==================== 2 failed, 11 passed in 23.04 seconds =====================
```

**Comment by nicoddemus** [comment]
That's weird... :worried: 

Perhaps using a remote debugger in one of the failing tests (`test_snt` for example) and execute it step by step to see if you can at least learn which line is producing the error.

**Comment by rekcahpassyla** [comment]
The problem is.... it isn't always that test. Which ones fail on each run appear to be random.

**Comment by rekcahpassyla** [comment]
The failing test has this format which is pretty much no logic in the test function other than the assert. 

```python
class SomethingRegression(TestCase):
    # Simple regression test for a known series
    a_expected = something
    b_expected = somethingelse

    @classmethod
    def setup_class(cls):
        # Calibration is done only once, with individual tests
        # checking values
        with open('{loc}/data.pickle'.format(loc=cwd), 'rb') as f:
            s = load(f)
        a, b, c = fit_some_values(s)
        cls.a = a
        cls.b = b
        cls.c = c

    def test_snt(self):
        assert_equal(self.a, self.a_expected, rtol=1e-7)
        assert_equal(self.b, self.b_expected, rtol=1e-7)
```

**Comment by nicoddemus** [comment]
> The problem is.... it isn't always that test. Which ones fail on each run appear to be random.

I see... that certainly makes it harder. :/

Looking at the code, keep in mind that each worker will execute `setup_class` once before running the tests from `SomethingRegression`. While reading the `data.pickle` from different processes in parallel should be OK, perhaps `fit_some_values` might be a problem.

I would try to change from `setup_class` to `setup_method` (and change the logic accordingly) and see what happens.

**Comment by rekcahpassyla** [comment]
No joy...  it now looks like this and I still have the same crashing behaviour

```python
class SomethingRegression(TestCase):
    # Simple regression test for a known series
    a_expected = something
    b_expected = somethingelse

    def setup_method(self, *args, **kwargs):
        # Calibration is done only once, with individual tests
        # checking values
        with open('{loc}/data.pickle'.format(loc=cwd), 'rb') as f:
            s = load(f)
        a, b, c = fit_some_values(s)
        self.a = a
        self.b = b
        self.c = c

    def test_snt(self):
        assert_equal(self.a, self.a_expected, rtol=1e-7)
        assert_equal(self.b, self.b_expected, rtol=1e-7)
```

**Comment by nicoddemus** [comment]
And if you don't call `fit_some_values` anymore? Use a dummy function which returns any `a, b, c`, just to see if `fit_some_values` might have something to do with it.

Perhaps could you come up with a reproducible case that you could share? I often find that this in the end helps find the actual problem.

**Comment by rekcahpassyla** [comment]
Thanks... I shall see what I can do.

**Comment by nicoddemus** [comment]
I see you are doing some regression testing, in this case you might take a look at [pytest-regtest](https://pypi.python.org/pypi/pytest-regtest).

**Comment by rekcahpassyla** [comment]
OK! I have something reproducible, and, I think it's got to do with statsmodels. This contrived example was ripped off from [this statsmodels examples](http://statsmodels.sourceforge.net/devel/examples/notebooks/generated/tsa_arma_1.html)

`py.test -v -r a -n 4 -m foo -s --max-slave-restart=0 ` with the following, will demonstrate the behaviour that 3 tests will fail with the `not properly terminated` error (which 3 is random). The number of failures with crashes is always `n-1`. 

`statsmodels` is not your problem! However, I now have a different question -- Is it possible to mark certain tests as never run in parallel? I know how to mark tests, but I'd like to know if it is possible to run some as parallel, and some as non-parallel, from the same command line. 

```python
@pytest.mark.foo
class TestFoo(TestCase):

    @classmethod
    def setup_class(cls, *args, **kwargs):
        import numpy as np
        import statsmodels.api as sm
        import pandas as pd
        from statsmodels.tsa.arima_process import arma_generate_sample
        np.random.seed(12345)
        arparams = np.array([.75, -.25])
        maparams = np.array([.65, .35])
        arparams = np.r_[1, -arparams]
        maparam = np.r_[1, maparams]
        nobs = 250
        y = arma_generate_sample(arparams, maparams, nobs)

        dates = sm.tsa.datetools.dates_from_range('1980m1', length=nobs)
        y = pd.TimeSeries(y, index=dates)
        arma_mod = sm.tsa.ARMA(y, order=(2,2))
        cls.arma_res = arma_mod.fit(trend='nc', disp=-1)

    def test_0(self):
        assert True

    def test_1(self):
        assert True

    def test_2(self):
        assert True

    def test_3(self):
        assert True

    def test_4(self):
        assert True

    def test_5(self):
        assert True

    def test_6(self):
        assert True

    def test_7(self):
        assert True

    def test_8(self):
        assert True
```

**Comment by nicoddemus** [comment]
> However, I now have a different question -- Is it possible to mark certain tests as never run in parallel?

There's nothing builtin for that, unfortunately. We will probably discuss something to the effect in the upcoming [pytest sprint]() though. 

We have the same problem here at work, specially for some sensitive GUI tests (@gabrielcnr will know what I'm talking about), and we solve it like this:

1. Mark tests which can't execute in parallel with some mark, say `serial`:

  ```python
  @pytest.mark.serial
  def test_snt(self):
        assert True
  ```

2. Execute pytest twice: once executing in parallel but **excluding** those with the mark, and then executing it again serially but executing only the tests **with** the mark:
  
  ```
  $ py.test -n4 -m "not serial"
  ...
  $ py.test -m serial
  ```

It works well for us. :grin:

**Comment by nicoddemus** [comment]
Oops, hit "send" too soon.

I will see if I can reproduce the issue, thanks for posting the example.

**Comment by rekcahpassyla** [comment]
Many thanks, that was what I was thinking of, it could work for us because we already have a setup whereby we start 3 different nose processes from our CI test slaves.

**Comment by nicoddemus** [comment]
@rekcahpassyla is still an issue? Also since then we have new distribution options which can be used to control how tests are sent to workers.

**Comment by rekcahpassyla** [comment]
Apologies for delay in replying- I actually stopped trying, because after I removed all the dodgy tests that were causing these issues, running with many test slaves was slower than our "poor man's concurrency" of just segmenting the tests ourselves. I haven't yet had time since Jun 10 2016 to look at trying again! We would still like to use this functionality but we're a bit swamped..

**Comment by nicoddemus** [comment]
No problem, thanks for the feedback. I will close this for now, feel free to open a new issue when/if you still encounter this. 😁

**Comment by cauvery** [comment]
I just encountered this issue today.  I am running the tests from Jenkins and the jenkins job is stuck with the last set to parallel tests to be executed.  And since the job is just hung when I abort it, it is giving 

[gw0] node down: Not properly terminated

Not sure what is causing this,  It was all working fine until yesterday.  I will try and see if I can get more details or steps to reproduce.

Thank you,
Cauvery.
Cauvery.

**Comment by ncclementi** [comment]
We are seeing this problem (on ubuntu CI though) quite a lot, over https://github.com/coiled/coiled-runtime/issues/404
it's not clear what the problem is, any input would be appreciated.
